fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeCollection.java

```diff
@@ -0,0 +1,172 @@
+        }
+    }
+
+    public void grant(short type, ActionSet actionSet, List<PEntryObject> objects, boolean isGrant) {
+        typeToPrivilegeEntryList.computeIfAbsent(type, k -> new ArrayList<>());
+        List<PrivilegeEntry> privilegeEntryList = typeToPrivilegeEntryList.get(type);
+        for (PEntryObject object : objects) {
+            PrivilegeEntry entry = findEntry(privilegeEntryList, object, isGrant);
+            PrivilegeEntry oppositeEntry = findEntry(privilegeEntryList, object, !isGrant);
+            if (oppositeEntry == null) {
+                // intend to grant with grant option, and there's no matching entry that grant without grant option
+                // or intend to grant without grant option, and there's no matching entry that grant with grant option
+                // either way it's simpler
+                addAction(privilegeEntryList, entry, actionSet, object, isGrant);
+            } else {
+                if (isGrant) {
+                    // intend to grant with grant option, and there's already an entry that grant without grant option
+                    // we should remove the entry and create a new one or added to the matching one
+                    removeAction(privilegeEntryList, oppositeEntry, actionSet);
+                    addAction(privilegeEntryList, entry, actionSet, object, isGrant);
+                } else {
+                    // intend to grant without grant option, and there's already an entry that grant with grant option
+                    // we should check for each action, for those that's not in the existing entry
+                    // we should create a new entry or added to the matching one
+                    ActionSet remaining = oppositeEntry.actionSet.difference(actionSet);
+                    if (! remaining.isEmpty()) {
+                        addAction(privilegeEntryList, entry, remaining, object, isGrant);
+                    }
+                }
+            }
+        } // for object in objects
+    }
+
+    public void revoke(short type, ActionSet actionSet, List<PEntryObject> objects, boolean isGrant) {
+        if (!typeToPrivilegeEntryList.containsKey(type)) {
+            LOG.debug("revoke a non-existence type {}", type);
+            return;
+        }
+        List<PrivilegeEntry> privilegeEntryList = typeToPrivilegeEntryList.get(type);
+        for (PEntryObject object : objects) {
+            PrivilegeEntry entry = findEntry(privilegeEntryList, object, isGrant);
+            if (entry != null) {
+                removeAction(privilegeEntryList, entry, actionSet);
+            }
+            // some of the actions may not granted
+            entry = findEntry(privilegeEntryList, object, !isGrant);
+            if (entry != null) {
+                // 1. intend to revoke with grant option but already grant object without grant option
+                // 2. intend to revoke without grant option but already grant object with grant option
```

Comment:
for case 2, is it resonable to remove the action, shall we need the user to specify with grant option precisely here?

Reply:
Make sense to me, and I have also confirmed with our PM, this behavior is as expected.

Reply:
ok